core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/refresh/extension/InteractionRefreshExtension.java

```diff
@@ -0,0 +1,188 @@
+   */
+  private static Map<EObject, Collection<EObject>> directEventsCache = new ConcurrentHashMap<>();
+
+  /**
+   * Cache for InstanceRole -> EventContext structure.
+   */
+  private static Map<InstanceRole, List<EventContext>> instanceRoleToEventContextsCache = new ConcurrentHashMap<>();
+
+  /**
+   * Is InteractionRefreshExtension caches enabled ?
+   */
+  private static boolean isCacheEnabled;
+
+  /**
+   * @see org.eclipse.sirius.business.api.refresh.IRefreshExtension#beforeRefresh(org.eclipse.sirius.DDiagram)
+   */
+  @Override
+  public void beforeRefresh(DDiagram diagram) {
+      enableRefreshCache();
+  }
+
+  /**
+   * @see org.eclipse.sirius.business.api.refresh.IRefreshExtension#postRefresh(org.eclipse.sirius.DDiagram)
+   */
+  @Override
+  public void postRefresh(DDiagram diagram) {
+    clearCaches();
+    disableRefreshCache();
+  }
+
+  private void clearCaches() {
+    interactionCache.clearCache();
+    directEventsCache.clear();
+    instanceRoleToEventContextsCache.clear();
+    interactionFragmentToTimeLapseCache.clear();
+  }
+
+  /**
+   * 
+   * @param function
+   * @param parameter
+   * @return If enabled, return the cached result if any or apply the function to the given parameter and cache the
+   *         result before returning it. //
+   */
+  public static <P, R> R getInteractionCache(Function<P, R> function, P parameter) {
+    return interactionCache.get(function, parameter);
+  }
+
+  /**
+   * Get EventContext structure for instanceRole
+   * 
+   * @param instanceRole
+   *          InstanceRole
+   * @return List<EventContext>
+   */
+  public static List<EventContext> getInstanceRoleToEventContextCache(InstanceRole instanceRole) {
+    return instanceRoleToEventContextsCache.get(instanceRole);
+  }
+
+  /**
+   * Put EventContext structure for instanceRole in cache.
+   * 
+   * @param instanceRole
+   *          InstanceRole
+   * @param structure
+   *          List<EventContext>
+   */
+  public static void putInstanceRoleToEventContextsCache(InstanceRole instanceRole, List<EventContext> structure) {
+    instanceRoleToEventContextsCache.put(instanceRole, structure);
+  }
+
+  /**
+   * Get element container.
+   * 
+   * @param capellaElement
+   *          EObject
+   * @return Collection<EObject>
+   */
+  public static Collection<EObject> getDirectEventsFromCache(EObject capellaElement) {
+    return directEventsCache.get(capellaElement);
+  }
+
+  /**
+   * Put element -> container in cache.
+   * 
+   * @param element
+   *          EObject
+   * @param container
+   *          Collection<EObject>
+   */
+  public static void putDirectEventsInCache(EObject element, Collection<EObject> container) {
+    directEventsCache.put(element, container);
+  }
+
+  /**
+   * @param key
+   *          InteractionFragment
+   * @return Event corresponding to key.
+   */
+  public static Optional<TimeLapse> getTimeLapseFromCache(InteractionFragment key) {
```

Comment:
Why does this method return an optional? All of the other methods return null if the value is absent. 
Let's return a `TimeLapse` here ;)

Reply:
Because it is a special case :

 it is used  to simulate the "MixEnd" case : some MessageEnd are used as start/finish of an Execution additionnaly to being the start/finish of a message. But there is no end -> Execution/TimeLapse info in the M2.  
We also use this cache to retrieve the InteractionState > StateFragment information.
And to be consistent, we also store ExecutionEnd -> Execution. 

If the cache returns null, IMO, this means that we put no entry in it. If the cache returns en empty Optional, this means that the TimeLapse in the model has no start or finish. 

But looking at the implementation, it seems that putTimeLapseInCache should do nothing if the provided key is null and you are right getTimeLapseFromCache could return a TimeLapse, without Optional.